tr2: refactor sound effect loading

lahm86 opened the pull request:
Resolves #2615.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This adjusts the reading of `main.sfx` to allow support for offsets that aren't necessarily in order, such as when we later come to inject sound fixes. We do this by scanning the file once and noting positions and lengths of all SFX, then scan over the defined offset to select the correct data.

The game flow has been updated too to support the default SFX file being defined there, as well as allowing a different one to be used in each level. This can be tested using the Golden Mask or TR3 equivalent files (although the sounds won't make sense 😄).

lahm86 committed:
tr2/level: refactor sound effect loading

This refactors the handling of main.sfx reading, to allow samples to be
loaded in any order of offset definition, and hence to support sound
effect injection.

rr- commented on src/tr2/game/level.c:
```diff
@@ -31,66 +31,85 @@
     const char *full_path = File_GetFullPath(file_name);
     LOG_DEBUG("Loading samples from %s", full_path);
-    MYFILE *const fp = File_Open(full_path, FILE_OPEN_READ);
+    VFILE *const fp = VFile_CreateFromPath(full_path);
```
While I know that VFile has attractive APIs, this is sketchy, as it'll load the entire file into memory at once as VFile doesn't know how to stream internally. It seems to cause a 50% performance downgrade for this function, going from 2 ms to 3+ ms on each level load.

rr- commented on src/tr2/game/level.c:
```diff
@@ -31,66 +31,85 @@
-        memcpy(sample_data, header, 0x2C);
-        File_ReadData(fp, sample_data + 0x2C, aligned_size);
+        const WAV_ENTRY entry = *(const WAV_ENTRY *)Vector_Get(entries, offset);
+        char *sample_data = Memory_Alloc(entry.size);
```
We could prevent double allocation for each sample by exposing the VFile's internal buffer. While this might improve performance, it does create some issues, as it's not the prettiest design, and we'd still end up reading all samples, not just those necessary for the level, so the benefits are uncertain.

rr- replied:
Actually it's a triple allocation, as `Audio_Sample_LoadSingle` copies the data too :D

rr- reviewed (commented):
I am experiencing a crash when exiting the game related to freeing the `sfx_path`:
```
#6  0x00007ffff7cc69ab in free () from /usr/lib/libc.so.6
#7  0x0000555555738944 in Memory_Free (memory=0x55555888fd60) at ../../../src/tr2/subprojects/libtrx/memory.c:43
#8  0x00005555557389cd in Memory_FreePointer (arg=0x55555653f8d0 <g_GameFlow+144>) at ../../../src/tr2/subprojects/libtrx/memory.c:53
#9  0x00005555556fbbea in GF_Shutdown () at ../../../src/tr2/subprojects/libtrx/game/game_flow/common.c:97
#10 0x00005555556db0a9 in Shell_Shutdown () at game/shell/common.c:480
#11 0x000055555571d855 in Shell_Terminate (exit_code=0) at ../../../src/tr2/subprojects/libtrx/game/shell/common.c:100
#12 0x000055555571dc60 in main (argc=1, argv=0x7fffffffd588) at ../../../src/tr2/subprojects/libtrx/game/shell/main.c:31
```
Most likely this is caused by the default settings holding a constant buffer, which gets repurposed as dynamically allocated strings rather on. To combat this, I'd leave it as `nullptr` and create a `#define` with the default path at the usage site in the event the string wasn't provided by the gameflow editor.

Personally, I think it's worth sticking to plain `File` over `VFile`.

lahm86 commented:
I've reverted back to using `FILE` and have defaulted the SFX path to `nullptr`.

I think there will still be a minor performance hit because we are caching the offsets, so inevitably that additional loop involves some extra time. On my end, it costs an extra ~0.8ms compared with develop. I can't really see a way to optimise this further, but would appreciate any further thoughts. We're being careful in the first loop to only cache the offsets so the footprint there is minimal, it's just unfortunate that we need to do so for every sample in the game.

lahm86 committed:
tr2/game_flow: specify SFX file in game flow

This removes the hardcoded reference to main.sfx, and allows it to be
defined globally in the game flow, and per level.

Resolves #2615.

lahm86 commented:
Thank you @rr- for providing a much neater solution to the SFX loading approach 😌
